sdk/python/kfp/v2/dsl/component_bridge.py

```diff
@@ -0,0 +1,199 @@
+          'Incompatible argument passed to the input "{}" of component "{}": '
+          .format(input_name, component_spec.name))
+
+      arguments[input_name] = str(argument_value)
+
+      if type_utils.is_artifact_type(input_type):
+        # argument_value.op_name could be none, in which case an importer node
+        # will be inserted later. Use output_artifact_key to preserve the name
+        # of pipeline parameter which is needed by importer.
+        pipeline_task_spec.inputs.artifacts[input_name].producer_task = (
+            argument_value.op_name or '')
+        pipeline_task_spec.inputs.artifacts[input_name].output_artifact_key = (
+            argument_value.name)
+      elif type_utils.is_parameter_type(input_type):
+        pipeline_task_spec.inputs.parameters[
+            input_name].runtime_value.runtime_parameter = argument_value.name
+      else:
+        raise NotImplementedError(
+            'Unsupported input type: "{}"'.format(input_type))
+    elif isinstance(argument_value, str):
+      pipeline_task_spec.inputs.parameters[
+          input_name].runtime_value.constant_value.string_value = argument_value
+    elif isinstance(argument_value, int):
+      pipeline_task_spec.inputs.parameters[
+          input_name].runtime_value.constant_value.int_value = argument_value
+    elif isinstance(argument_value, float):
+      pipeline_task_spec.inputs.parameters[
+          input_name].runtime_value.constant_value.double_value = argument_value
+    else:
+      raise NotImplementedError(
+          'Input argument supports only the following types: PipelineParam'
+          ', str, int, float. Got: "{}".'.format(argument_value))
+
+  for output in component_spec.outputs or []:
+    normalized_output_type = output.type.lower()
+    if type_utils.is_artifact_type(output.type):
+      pipeline_task_spec.outputs.artifacts[
+          output
+          .name].artifact_type.schema_title = type_utils.get_artifact_type(
+              output.type)
+    elif type_utils.is_parameter_type(output.type):
+      pipeline_task_spec.outputs.parameters[
+          output.name].type = type_utils.get_parameter_type(output.type)
+    else:
+      raise NotImplementedError('Unsupported output type: "{}"'.format(
+          output.type))
+
+  def _input_artifact_placeholder(input_key: str) -> str:
+    return "{{{{$.inputs.artifacts['{}'].path}}}}".format(input_key)
+
+  def _input_parameter_placeholder(input_key: str) -> str:
+    return "{{{{$.inputs.parameters['{}']}}}}".format(input_key)
+
+  def _output_artifact_placeholder(input_key: str) -> str:
+    return "{{{{$.outputs.artifacts['{}'].path}}}}".format(input_key)
+
+  # IR placeholders are decided merely based on the declared type of the input.
+  # It doesn't matter wether it's InputValuePlaceholder or InputPathPlaceholder
+  # from component_spec.
+  placeholder_arguments = {
+      input_spec.name: _input_artifact_placeholder(input_spec.name)
```

Comment:
>_input_artifact_placeholder

I'm not sure this will work. These arguments will be used to substitute the `inputValue` placeholder while `_input_artifact_placeholder` is a placeholder for a path. Getting path instead of value is unexpected.

You'll need a placeholder that resolves to a value here.

Reply:
I should add more comments here. I think it works because of the importer node auto inserted by the compiler, the inputValue will be feed into importer as value (parameter), while the output of importer node is a path. Does that make sense? Can you please help review the golden json output that may explain this?

Reply:
I've looked at the golden json and it looks OK.
However I'm still not 100% sure the logic on the next line is fully correct.
It works great for URI inputs, but for other artifact types it looks pretty confusing. The code seems to imply that a Model **is** a URI, but usually a model is a binary file with stored weights and network architecture.

Perhaps instead of
```python
if type_utils.is_artifact_type(input_spec.type)
```
it should be
```python
if type_utils.is_uri_type(input_spec.type)
```

It's OK not to make a change now, since the code is experimental. But we might want to think about making this less confusing.


Currently in KFP there are two styles of consuming bigger data - through files or through URIs (for example, all [KFP TFX components](https://github.com/kubeflow/pipelines/tree/master/components/deprecated/tfx) have those two flavors):

In this example, the component expects a model.
```yaml
inputs:
- {name: Model, type: Model}
implementation:
  container:
    command:
    - program.py
    - --model-path
    - {inputPath: Model}
```

In this example it's very clear that the input is a URI, not a model itself.
```yaml
inputs:
- {name: Model URI, type: {Uri: {data_type: Model}}}
implementation:
  container:
    command:
    - program.py
    - --model-uri
    - {inputValue: Model URI}
```